File: leetcode-75/09-binary-tree-dfs/34-leaf-similar-trees/solution.py
"""
LeetCode 872. Leaf-Similar Trees
https://leetcode.com/problems/leaf-similar-trees/

Approach: DFS each tree (left before right) to collect its leaf
values in left-to-right order, then compare the two sequences.

Time:  O(n + m) - every node of both trees is visited once
Space: O(n + m) for the leaf lists, plus O(h1 + h2) recursion stack
"""

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def get_leaves(node: Optional[TreeNode], leaves: List[int]) -> None:
            if node is None:
                return
            if node.left is None and node.right is None:
                leaves.append(node.val)
                return
            get_leaves(node.left, leaves)
            get_leaves(node.right, leaves)

        leaves1: List[int] = []
        leaves2: List[int] = []
        get_leaves(root1, leaves1)
        get_leaves(root2, leaves2)

        return leaves1 == leaves2


def assert_equal(actual, expected, label):
    if actual != expected:
        print(f"FAIL [{label}]: got {actual}, expected {expected}")
    else:
        print(f"PASS [{label}]")


if __name__ == "__main__":
    sol = Solution()
    t1 = TreeNode(1, TreeNode(2), TreeNode(3))
    t2 = TreeNode(1, TreeNode(3), TreeNode(2))
    print(sol.leafSimilar(t1, t1))  # True (same tree)
    print(sol.leafSimilar(t1, t2))  # False

    # Edge cases
    assert_equal(sol.leafSimilar(TreeNode(5), TreeNode(5)), True, "single node, same value")
    assert_equal(sol.leafSimilar(TreeNode(5), TreeNode(6)), False, "single node, different value")

    # Different shapes, same leaf sequence [2, 3]:
    shape_a = TreeNode(1, TreeNode(2), TreeNode(3))
    shape_b = TreeNode(1, TreeNode(10, TreeNode(2)), TreeNode(3))
    assert_equal(sol.leafSimilar(shape_a, shape_b), True, "different shapes, same leaf sequence")

    assert_equal(
        sol.leafSimilar(
            TreeNode(5, TreeNode(6), TreeNode(7)),
            TreeNode(5, TreeNode(7), TreeNode(6)),
        ),
        False,
        "same values, mirrored leaf order",
    )

File: leetcode-75/09-binary-tree-dfs/34-leaf-similar-trees/solution.ts
/**
 * LeetCode 872. Leaf-Similar Trees
 * https://leetcode.com/problems/leaf-similar-trees/
 *
 * Approach: DFS each tree (left before right) to collect its leaf
 * values in left-to-right order, then compare the two sequences.
 *
 * Time:  O(n + m) - every node of both trees is visited once
 * Space: O(n + m) for the leaf lists, plus O(h1 + h2) recursion stack
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function getLeaves(node: TreeNode | null, leaves: number[]): void {
  if (node === null) return;

  if (node.left === null && node.right === null) {
    leaves.push(node.val);
    return;
  }

  getLeaves(node.left, leaves);
  getLeaves(node.right, leaves);
}

function leafSimilar(root1: TreeNode | null, root2: TreeNode | null): boolean {
  const leaves1: number[] = [];
  const leaves2: number[] = [];

  getLeaves(root1, leaves1);
  getLeaves(root2, leaves2);

  if (leaves1.length !== leaves2.length) return false;

  for (let i = 0; i < leaves1.length; i++) {
    if (leaves1[i] !== leaves2[i]) return false;
  }

  return true;
}
function assertEqual(actual: unknown, expected: unknown, label: string): void {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) {
    console.error(`FAIL [${label}]: got ${a}, expected ${e}`);
  } else {
    console.log(`PASS [${label}]`);
  }
}

if (require.main === module) {
  // Example usage:
  const t1 = new TreeNode(3, new TreeNode(5, new TreeNode(6), new TreeNode(2, new TreeNode(7), new TreeNode(4))), new TreeNode(1, new TreeNode(9), new TreeNode(8)));
  const t2 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
  console.log(leafSimilar(t1, t1)); // true (same tree)
  console.log(leafSimilar(t2, new TreeNode(1, new TreeNode(3), new TreeNode(2)))); // false

  // Edge cases
  assertEqual(leafSimilar(new TreeNode(5), new TreeNode(5)), true, 'single node, same value');
  assertEqual(leafSimilar(new TreeNode(5), new TreeNode(6)), false, 'single node, different value');

  // Different shapes, same leaf sequence [2, 3]:
  // shapeA: 1 -> (left: leaf 2, right: leaf 3)
  const shapeA = new TreeNode(1, new TreeNode(2), new TreeNode(3));
  // shapeB: 1 -> (left: 10 -> left: leaf 2, right: leaf 3)
  const shapeB = new TreeNode(1, new TreeNode(10, new TreeNode(2)), new TreeNode(3));
  assertEqual(leafSimilar(shapeA, shapeB), true, 'different shapes, same leaf sequence');

  assertEqual(
    leafSimilar(new TreeNode(5, new TreeNode(6), new TreeNode(7)), new TreeNode(5, new TreeNode(7), new TreeNode(6))),
    false,
    'same values, mirrored leaf order'
  );
}

export { leafSimilar, TreeNode };
